### PR TITLE
feat(cosmosdb): Add Private endpoint support for Azure CosmosDB

### PR DESCRIPTION
**Pull request by jlogel90**
### Checklist:

* [ ] Have you added Release Notes in the docs repositories? -> Don't know what to do, not mentioned in CONTRIBUTING.md
* [X] Have you followed the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)?

make_run-integration-tests_results

### CODE REVIEW

**Comment by jhvhs** [comment]
@jlogel90 Thank you very much for your pull request. We should be able to review it shortly.

Don't worry about the documentation - it will be managed separately.

**Comment by zucchinidev** [comment]
Hello,

Thank you very much for the PR, @jlogel90.

We are creating a couple of examples with integration and Terraform tests so you can use them as a reference and add the necessary tests to check your changes. Once I add the examples I will tell you in general how to proceed to carry out the new changes and that you can launch the tests.

Meanwhile, tell you that one of the GitHub actions failed to analyze the code. The error is due to the format. To fix it, simply run the following command on the project root:

```
make format
```

After executing the command, check the modified files, commit and push so that the GitHub actions are launched again.

If you have any questions, don’t hesitate to write.

**Comment by jlogel90** [comment]
> Hello,
> 
> Thank you very much for the PR, @jlogel90.
> 
> We are creating a couple of examples with integration and Terraform tests so you can use them as a reference and add the necessary tests to check your changes. Once I add the examples I will tell you in general how to proceed to carry out the new changes and that you can launch the tests.
> 
> Meanwhile, tell you that one of the GitHub actions failed to analyze the code. The error is due to the format. To fix it, simply run the following command on the project root:
> 
> ```
> make format
> ```
> 
> After executing the command, check the modified files, commit and push so that the GitHub actions are launched again.
> 
> If you have any questions, don’t hesitate to write.

Hello,

Thanks for taking time to review my PR.
I did run make format and commit result.

Regards.

**Comment by zucchinidev** [comment]
Hi @jlogel90,

On the contrary, many thanks to you for contributing. Now all GitHub actions have been successfully executed. 

The tests we use to check the intermediate layer created from the properties defined in the YML files that correspond to the definition of the services, and that check the correct assignment of variables when performing provision or update operations are on their way. When these new tests are finished, I will ask you to add assertions for the two properties you have added, in other words, the `private_endpoint_subnet_id` and `private_dns_zone_ids` properties.

I am also preparing the tests that check the logic applied to our Terraform code. With these Terraform tests, you can verify the valid values of the `resource_group` and `private_endpoint_enabled` properties, as well as, the correct configuration of the new resource you added `azurerm_private_endpoint`. I will keep you in the loop with the progress.

File: terraform/azure-cosmosdb/provision-cosmosdb-sql.tf [hunk 1]
```diff
@@ -53,9 +55,9 @@ provider "azurerm" {
 }
 
 locals {
-  resource_group = length(var.resource_group) == 0 ? format("rg-%s", var.instance_name) : var.resource_group
-
+  resource_group                = length(var.resource_group) == 0 ? format("rg-%s", var.instance_name) : var.resource_group
   enable_virtual_network_filter = (var.authorized_network != "")
+  private_endpoint_enabled      = var.private_endpoint_subnet_id == null ? false : length(var.private_endpoint_subnet_id) > 0 ? true : false
```
Comment (zucchinidev) [inline terraform/azure-cosmosdb/provision-cosmosdb-sql.tf:60]: thoughts: you might want to simplify the logic to improve the readability

```
private_endpoint_enabled = var.private_endpoint_subnet_id == null || length(var.private_endpoint_subnet_id) == 0 ? false : true
```

File: terraform/azure-cosmosdb/provision-cosmosdb-sql.tf [hunk 2]
```diff
@@ -117,6 +119,30 @@ resource "azurerm_cosmosdb_sql_database" "db" {
   }
 }
 
+resource "azurerm_private_endpoint" "private_endpoint" {
+  name                = "${var.instance_name}-private_endpoint"
+  location            = var.location
+  resource_group_name = var.resource_group
+  subnet_id           = var.private_endpoint_subnet_id
+  tags                = var.labels
+  count               = local.private_endpoint_enabled ? 1 : 0
```
Comment (zucchinidev) [inline terraform/azure-cosmosdb/provision-cosmosdb-sql.tf:128]: praise: thanks for formatting the code

**Comment by zucchinidev** [comment]
Hi @jlogel90

We have already added the integration tests that will guide you to check the correct value assignment in the new properties you added `private_endpoint_subnet_id` and `private_dns_zone_ids`.

You can now add the corresponding assertions to verify that the provision or update operation accepts appropriate values for the properties mentioned above. This is the test file you should modify: [integration-tests/cosmosdb_sql_test.go](https://github.com/cloudfoundry/csb-brokerpak-azure/blob/main/integration-tests/cosmosdb_sql_test.go#L67). To run the tests you are going to modify you can run the following command in the project root:

```
make run-integration-tests
```

Any doubt you might have regarding this point, don't hesitate to contact me and I will be more than happy to answer you.

**Comment by zucchinidev** [comment]
Hello @jlogel90,

We have added the `Terraform` test suite so you can check that the added resource `azurerm_private_endpoint` is created with the desired values.

You can complete the test in this [file](https://github.com/cloudfoundry/csb-brokerpak-azure/blob/main/terraform-tests/cosmosdb_sql_test.go).

In addition, you can see a more detailed explanation of how to run the tests in this [tutorial](https://github.com/cloudfoundry/cloud-service-broker/blob/main/docs/user-guides/extending-a-service.md).

Finally, if you have any doubts, you can ask me and take a look at the tests that exist in this repository for a broader view of how to add assertions through [examples](https://github.com/cloudfoundry/csb-brokerpak-aws/tree/main/terraform-tests). 

Regards,

Andrea.

**Review by zucchinidev — CHANGES_REQUESTED** [review]

File: terraform/azure-cosmosdb/provision-cosmosdb-sql.tf [hunk 3]
```diff
@@ -117,6 +119,30 @@ resource "azurerm_cosmosdb_sql_database" "db" {
...
+
+  private_service_connection {
+    name                           = "${var.instance_name}-private_service_connection"
+    private_connection_resource_id = azurerm_cosmosdb_account.cosmosdb-account.id
+    subresource_names              = ["MongoDB"]
```
[the remark this answers is not in the document]
Comment (jlogel90) [reply terraform/azure-cosmosdb/provision-cosmosdb-sql.tf]: Hello, you're right. I fixed it. It was a left over from the PR I mainly reused https://github.com/cloudfoundry/csb-brokerpak-azure/pull/103

File: terraform/azure-cosmosdb/provision-cosmosdb-sql.tf (hunk 1 above)
Comment (jlogel90) [reply terraform/azure-cosmosdb/provision-cosmosdb-sql.tf]: Hello, it fixed though I' m not sure to understand syntax as I can' t find a related example in Terraform documentation

**Comment by jlogel90** [comment]
> Hello,
> 
> Thank you very much for the PR, @jlogel90.
> 
> We are creating a couple of examples with integration and Terraform tests so you can use them as a reference and add the necessary tests to check your changes. Once I add the examples I will tell you in general how to proceed to carry out the new changes and that you can launch the tests.
> 
> Meanwhile, tell you that one of the GitHub actions failed to analyze the code. The error is due to the format. To fix it, simply run the following command on the project root:
> 
> ```
> make format
> ```
> 
> After executing the command, check the modified files, commit and push so that the GitHub actions are launched again.
> 
> If you have any questions, don’t hesitate to write.

Hello,

Thanks for taking time to review my PR.
I did run make format and commit result.

Regards.

> Hi @jlogel90
> 
> We have already added the integration tests that will guide you to check the correct value assignment in the new properties you added `private_endpoint_subnet_id` and `private_dns_zone_ids`.
> 
> You can now add the corresponding assertions to verify that the provision or update operation accepts appropriate values for the properties mentioned above. This is the test file you should modify: [integration-tests/cosmosdb_sql_test.go](https://github.com/cloudfoundry/csb-brokerpak-azure/blob/main/integration-tests/cosmosdb_sql_test.go#L67). To run the tests you are going to modify you can run the following command in the project root:
> 
> ```
> make run-integration-tests
> ```
> 
> Any doubt you might have regarding this point, don't hesitate to contact me and I will be more than happy to answer you.

Hello, I kind of struggle on this, I'm not a developer and not familiar with go.
I figured it out that I should probably add something like this : 
`HaveKeyWithValue("private_endpoint_subnet_id", ?()),
 HaveKeyWithValue("private_dns_zone_ids", ?()),`

I looked here: 
https://onsi.github.io/gomega/#havekeywithvaluekey-interface-value-interface and 
https://onsi.github.io/gomega/#working-with-collections
But I don' t know what kind of "matcher" I should use: first variable is required and the other one is optional so could be empty and both of them are string types.

Thanks

File: terraform/azure-cosmosdb/provision-cosmosdb-sql.tf (hunk 1 above)
Comment (zucchinidev) [reply terraform/azure-cosmosdb/provision-cosmosdb-sql.tf]: Thank you very much @jlogel90. No worries, the logic applied in ternary operators is usually quite tricky. The logic was correct, but it is easier to read after this simplification

File: terraform/azure-cosmosdb/provision-cosmosdb-sql.tf (hunk 3 above)
[the remark this answers is not in the document]
Comment (zucchinidev) [reply terraform/azure-cosmosdb/provision-cosmosdb-sql.tf]: @jlogel90 Thanks a million!

**Comment by zucchinidev** [comment]
Hi @jlogel90,

First of all, many thanks for your contribution! :tada:

We understand that you are not familiar with tests or Go. Don’t worry. 

We will proceed to approve the changes and add the missing tests today :rocket:. Therefore, you can analyse them and get a more accurate idea of what was needed.

To facilitate the contribution process and encourage you to continue contributing, we will continue adding examples of tests you can use as a reference in future contributions. We intend to continue to improve the documentation and even record a video with recommendations to help in this process, so stay tuned. I am sure that you will soon be able to contribute without hesitation.

We expect new contributions, and always remember that it is a pleasure for us to help you, so never hesitate to ask.

Regards,

Andrea.

**Comment by zucchinidev** [comment]
Hi @jlogel90

The tests have been added. I recommend you take a look at the new changes of these two files `integration-tests/cosmosdb_sql_test.go` and `terraform-tests/cosmosdb_sql_test.go`, so that you can get a more generalised context and be able to advance in your next contribution!!

All the best!

**Pull request merged**